src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/BatchParserWrapper.cs

```diff
@@ -82,15 +82,15 @@ private static List<BatchDefinition> ConvertToBatchDefinitionList(string content
                     }
 
                     // Generate the rest batch definitions
-                    for (int index = 1; index < count - 1; index++)
+                    for (int index = 1; index < count - 1 ; index++)
                     {
                         lineDifference = batchInfos[index + 1].startLine - batchInfos[index].startLine;
                         position = ReadLines(reader, lineDifference, endLine);
                         endLine = position.Item1;
                         endColumn = position.Item2;
                         offset = offsets[index];
-                        batchText = content.Substring(offset, batchInfos[index].length);
-                        startLine = batchInfos[index].startLine;
+                        batchText = batchInfos[index].batchText;
+                        startLine = batchInfos[index].startLine + 1; // this is done for first batch but was missing for following batches
```

Comment:
Be careful with the batch indices. We've had issues here where the batch starting lines and end lines don't match up. You can verify this by running any query and then clicking on the `Line xx` in the Query Results window. It should highlight the actual run batch.

Reply:
Yeah - I'm kind of worried about making changes here. When I was doing the query history stuff I found a lot of weirdness with how the numbers end up and so making changes here could easily break functionality elsewhere.

Reply:
Agreed. Make sure we test out on mac/Linux as I see some code that searches for "\n" in OnBatchParserExecutionFinished and adjusts batch length.

---
In reply to: [309457799](https://github.com/microsoft/sqltoolsservice/pull/839#discussion_r309457799) [](ancestors = 309457799)

Reply:
Sure I will test that way.  

The batch indecies were giving wrong result for sustitution - that is why I added the text to be sure.

Regarding line numbers - I tried comparing with DacFx but couldntfully since its changed much. In there we have no 0 based index so we use the startline as is. Here we have 0 based index and we add +1 to first line but ot to later ones. 

Those are the only two changes here. Let's discuss if the second one looks too risky.

Reply:
@abist update - validated selecting lines from message